figures.md: fixes and improvements

maxim-belkin opened the pull request:
1. Fix the script so that all figures are displayed.
2. Add image numbers.
3. Don't add headers for episodes without images.
4. Add horizontal rulers to better separate adjacent images.
   Don't add a ruler after the last image in the episode.
5. Add empty lines to improve code readability.

I'll add before and after screenshots later.

maxim-belkin commented:
![before_and_after-01](https://user-images.githubusercontent.com/13123663/42727027-0e3d119e-878e-11e8-9a5a-dcb5c6682620.png)

maxim-belkin commented:
Note that "Creating Another Variable" figure is missing in the "Before" shot.

maxim-belkin commented:
In fact, a lot of figures are missing when the current script is used

rgaiacs commented on bin/boilerplate/_extras/figures.md:
```diff
@@ -5,22 +5,55 @@ title: Figures
+
+            var htitle = document.createElement('h2');
+            htitle.innerHTML = 'Episode "' + h1text + '"';
+            htitle.style = "margin-top: 100px;";
```
Can we use a class instead of inline CSS?

maxim-belkin replied:
I don't know of a dedicated class for H2, H3, etc. in SWC/DC lessons (only for H1). I can create one, though, like `.figure-title`.

rgaiacs replied:
In `_extras/figures.md` you have

```
{% for episode in site.episodes %}
<article id="{{ episode.url }}"></article>
{% endfor %}
```

you can change it to

```
{% for episode in site.episodes %}
<article id="{{ episode.url }} class="figures"></article>
{% endfor %}
```

and add a rule to the CSS file like

```
article.figures h1,
article.figures h2,
article.figures h3,
article.figures h4,
article.figures h5,
article.figures h6 {
    ...
}
```

I think this approach is better because contributors to the lesson should have some knowledge of HTML to be able to read the file and get one idea of what it is doing but we can't expect that contributors will be confident in read Javascript.

rgaiacs commented on bin/boilerplate/_extras/figures.md:
```diff
@@ -5,22 +5,55 @@ title: Figures
+
+              var title = document.createElement('h3');
+              title.innerHTML = "Figure " + image_num + ". " + image.alt;
+              title.style = "margin-top: 20px;"
```
Again, can we use a class instead of inline CSS?

rgaiacs commented on bin/boilerplate/_extras/figures.md:
```diff
@@ -5,22 +5,55 @@ title: Figures
+
+              var img = document.createElement('img');
+              img.src = image.src;
+              img.style = "display: block; margin: 20px auto;"
```
Again, can we use a class instead of inline CSS?

rgaiacs commented on bin/boilerplate/_extras/figures.md:
```diff
@@ -5,22 +5,55 @@ title: Figures
+              article_here.appendChild(img);
+
+              if (image_num < images.length) {
+                var hr = document.createElement('hr');
```
I don't like to have many lines between the images.

maxim-belkin replied:
I added these horizontal lines (and extra margins) only after looking at the final page. Without them, the title of the next image might seem related to the previous image. I'm open to suggestions how to make figure--title connection more obvious.

One possibility: images like in science papers. That is, an image, and immediately below it `Figure 01: Blah-blah-blah`

rgaiacs commented:
> In fact, a lot of figures are missing when the current script is used

Can you provide some examples?

maxim-belkin commented:
> > In fact, a lot of figures are missing when the current script is used
>
> Can you provide some examples?

Sure. Have a look at episode 1 in python novice inflammation:
https://swcarpentry.github.io/python-novice-inflammation/01-numpy/index.html

and page with all figures:
https://swcarpentry.github.io/python-novice-inflammation/figures/index.html

There are 6 figures on the page with "all" figures, whereas the episode has 10 figures.

fmichonneau commented:
is this ready to merge? (the result looks good to me but I haven't had a chance to review the code submitted).

maxim-belkin commented:
I haven't implemented the class-thing that Raniere requested (it's easy, I just didn't have the time to do that)

maxim-belkin commented:
I think I found the solution should satisfy all of us: I removed the word "Episode" from the `h2` headers, removed `h3` headers, removed manually-added `hr` under `h2`s. 

With the exception for the special top-margin for `h2` headers, this page should pick up all the styles defined for the main lesson. The special top margin for `h2` headers should help better separate the "episodes" on the page.

I'm going to merge this PR and close #343 as all of the changes there have been added back here.
